contracts/blade/DestinationGateway.sol

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+import "./BaseBridgeGateway.sol";
+
+contract DestinationGateway is BaseBridgeGateway {
+    /// @custom:security write-protection="onlySystemCall()"
+    // slither-disable-next-line protected-vars
+    mapping(uint256 => bool) public processedEvents;
+
+    event BridgeMessageResult(uint256 indexed counter, bool indexed status, bytes message);
+
+    /**
+     * @notice receives the batch of messages and executes them
+     * @param batch batch of messages
+     * @param signature the aggregated signature submitted by the proposer
+     * @param bitmap bitmap of which validators signed the message
+     */
+    // slither-disable-next-line protected-vars
+    function receiveBatch(
+        BridgeMessageBatch calldata batch,
+        uint256[2] calldata signature,
+        bytes calldata bitmap
+    ) external {
+        _verifyBatch(batch);
+
+        bytes memory hash = abi.encode(keccak256(abi.encode(batch)));
+        verifySignature(bls.hashToPoint(DOMAIN, hash), signature, bitmap);
+
+        uint256 length = batch.messages.length;
+        for (uint256 i = 0; i < length; ) {
+            _executeBridgeMessage(batch.messages[i]);
+
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    /**
+     * @notice Internal function that verifies the batch
+     * @param batch batch to verify
+     */
+    function _verifyBatch(BridgeMessageBatch calldata batch) private view {
+        require(batch.messages.length > 0, "EMPTY_BATCH");
+
+        uint256 destinationChainId = block.chainid;
+        require(batch.destinationChainId == destinationChainId, "INVALID_DESTINATION_CHAIN_ID");
+        for (uint256 i = 0; i < batch.messages.length; ) {
+            BridgeMessage memory message = batch.messages[i];
+            require(message.sourceChainId == batch.sourceChainId, "INVALID_SOURCE_CHAIN_ID");
+            require(message.destinationChainId == destinationChainId, "INVALID_DESTINATION_CHAIN_ID");
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    function _executeBridgeMessage(BridgeMessage calldata message) private {
+        require(!processedEvents[message.id], "DestinationGateway: BRIDGE_MESSAGE_IS_PROCESSED");
+        // Skip transaction if client has added flag, or receiver has no code
+        if (message.receiver.code.length == 0) {
```

Comment:
Just an idea, but needs some more investigation...

It would be awesome if we could check whether `message.receiver` address is of `IStateReceiver` type.
Probably worth investigating it further (https://medium.com/quick-programming/what-is-erc165-and-why-you-should-use-it-d0641a2f29e5).